### ISSUE TITLE
Composer/working strategies

### ISSUE THREAD

**Issue by craigcarnell** [magento/magento2#2168]
At the moment, I have cloned the magento2 repo and I've started to do some work here on themes/modules etc.

What I'm trying to do is work out the best way for us to be able to pull upstream changes down while at the same time separate our themes/modules from Magento.

Naturally, I can create individual repos with a clone of magento2 and use composer within magento2 to fetch those down.

But I also saw that https://github.com/magento/magento2-community-edition/blob/master/composer.json exists.

Would it be more prudent for example, to use that composer.json as the root of my git repo, and pull down magento2ce, pull down my own shared custom modules using composer, but keeping relevant themes/modules unique to this site in the repo etc., thus keeping my own repo very clean of core magento2 code?

I will be deploying the code via capistrano and jenkins.

I plan on having multiple magento2 repos for different sites (that are not very common with each other) so you can see by doing this it will make patching etc a lot easier.

Or does this all sound like a bad idea and I should just clone magento2

I did find these instructions, but as things change so frequently, I am not sure if they are still relevant...

http://magenticians.com/installing-magento-2-composer
composer create-project magento/product-community-edition magento2 -s dev --prefer-dist

I imagine doing the later might be problematic if I want to checkout develop branch or some ordering of modules etc in my own repo

Edit: well after trying magento2-composer-installer it fetched it down just nicely but when I did my own customization to app/code/ and app/design/frontend/my_theme it was not picked up by git

**Comment by alankent** [comment]
We are experiencing the normal joy of getting all the last bits finished and brought together. There is something on the internal list of things to try that we have not got to. Would you like to experiment a bit and see if it works? (We have not got to it yet.)

At GA, we will have official releases of code in an official Composer repo. We don't have that today, so things feel half baked. They are right now, but they will be cleaned up by GA. It will be good... "trust me!"... (or not!)

If you look in https://github.com/magento/magento2/blob/develop/app/etc/NonComposerComponentRegistration.php you will find there are a list of directories we look for the registration.php files in. We were reviewing if we should allow say an environment variable to add to this directory, so you can add your own directories or glob patterns to this list. (Semicolon separated for Windows, colon separated for *nix.)

The idea is to allow an extension developer to create their own directory, put their code there, manage it by git etc. But it is separate from the rest of the site. We will support creating a site completely via composer, with the downloaded modules going into 'vendor'. But maybe you want to have 5 sites set up, corresponding to CE-2.0.0, CE-2.0.1, CE-2.0.2, and so on. You may *also* want to test your extension with the 'develop' branch (knowing it is still unstable). So for an extension developer, it feels nicer to be able to keep your extension code separate to the main site, to make code management easier. (This is not recommended for production - but it is useful to an extension developer.)

This would also give an extension developer more options in building their own test automation pipeline.

If someone wanted to experiment with the above and see how it feels, that feedback would be great. If it works (I cannot see why it would not), then we could look at putting a few extra lines in the standard file we ship to read the environment variable to extend the list of search directories, and then you can have the one extension shared by several Magento installations - created via composer or git clone.

We will try it internally - but if someone beats us to it then great! (One less thing for us to get to.) But nothing like customers to tell us whether it really helps or not!  To experiment, just manually edit the file to add your own directory in for now.

**Comment by alankent** [comment]
...and as soon as I go back to my email, I noticed there is a known bug as to when the code cannot be outside the Magento root directory just yet. Something about `bin/magento setup:static-content:deploy` not working in this case. So be aware there are known problems.

**Comment by craigcarnell** [comment]
@alankent Strangest thing is when I used magento composer installer repo, it installed the modules into vendors but the files also exist in app/code/Magento and not symlinks?

**Comment by alankent** [comment]
The old composer approach (started as a community Hackthon project by the way) got Composer & Magento to play ball by downloading to vendor then copying to app/code. Magento only picked it up under app/code. We have just recently got it so Magento can load from 'vendor' directly - but we need to keep things backwards compatible and some dependencies we found used "*", so there might be copying still going on at the moment. We are trying to clean this up, but there may be issues like this still.

We don't recommend doing git clone from the repo for production (especially 'develop'!) as patching is not guaranteed to work. Our upgrade scripts work from official path level to official patch level only.

"Clever people" can go to the git repo, but it will be on your head more to get it right. We are recommending everyone to use official downloads as they will be more controlled, more tested, easier to upgrade etc.  So if you want the slick, tested, robust experience then download official packages. If you want the latest code always, git clone.  (But its not quite ready yet - almost there!!!)
